aten/src/ATen/native/Copy.cpp

```diff
@@ -104,44 +106,20 @@ void _copy_same_type__cpu(Tensor& self, const Tensor& src) {
   if (self.is_same(src)) {
     return;
   }
-
-  // TODO: Replace this with TensorIterator!
-  bool serial_path = false;
   if (self.numel() == src.numel()) {
     if (self.is_contiguous() && src.is_contiguous()) {
-      copy_kernel(kCPU, self, src);
+      return copy_kernel(kCPU, self, src);
     } else if (copy_transpose_valid(self, src)) {
-      _copy_same_type_transpose_(self, src);
-    } else {
-#ifdef _OPENMP
-      if (!in_parallel_region()) {
-        AT_DISPATCH_ALL_TYPES_AND(
-          at::ScalarType::Half, self.scalar_type(), "_copy_same_type_", [&]() {
-            at::CPU_tensor_parallel_apply2<scalar_t, scalar_t>(
-                self, src, [](scalar_t& self_val, const scalar_t& src_val) {
-                  self_val = src_val;
-                });
-          });
-      } else {
-        serial_path = true;
-      }
-#else
-      serial_path = true;
-#endif
+      return _copy_same_type_transpose_(self, src);
     }
-  } else {
-    serial_path = true;
-  }
-
-  if (serial_path) {
-    AT_DISPATCH_ALL_TYPES_AND(
-      at::ScalarType::Half, self.scalar_type(), "_copy_same_type_", [&]() {
-        at::CPU_tensor_apply2<scalar_t, scalar_t>(
-            self, src, [](scalar_t& self_val, const scalar_t& src_val) {
-              self_val = src_val;
-            });
-        });
   }
+  auto iter = TensorIterator::unary_op(self, src, /* resize_outputs = */ false);
```

Comment:
The `unary_kernel` call should be pushed into CopyKernel.cpp and it should completely replace the current `copy_kernel`.

i.e. _copy_same_type__cpu should just be:
```
void _copy_same_type__cpu(Tensor& self, const Tensor& src) {
  if (self.is_same(src)) {
    return;
  }
  if (self.numel() == src.numel() && copy_transpose_valid(self, src)) {
    _copy_same_type_transpose_(self, src);
  }
  copy_kernel(kCPU, *iter);
}
```

You should also add an extra case to `unary_kernel` to handle the case where one operand is effectively a scalar (stride 0):

`is_contiguous_s1`

Here's the corresponding case for the binary_kernel:

https://github.com/pytorch/pytorch/blob/3e0b46b6d132c22af84769a6df1ef8d1f785dbe2/aten/src/ATen/native/cpu/Loops.h#L21-L26

`is_contiguous_s1` should be the same but no check of `strides[2]` (because there are only two operands)

Reply:
Will handle next #19437

Reply:
Please move to the cpu subdirectory before landing